### DISCUSSION TITLE
Saving responses from Textract

### DISCUSSION THREAD

**Discussion by DexterW** [soto-project/soto#663] [Q&A]
Hello -

I'm working with Textract and would like to save the raw json results to a file in order to avoid re-hitting the API if I've already called for a specific file.  `Textract.DetectDocumentTextResponse` is `Decodable` but not `Encodable`.  I'm trying to find the base way to save responses -- does anyone have any experience doing this?  Would it be worth making a PR to make the response `Encodable`?

**Comment by adam-fowler** [comment]
Adding `Encodable` conformance to the response shapes would add a lot of code to the libraries. The soto libs are already large enough. You can add the `Encodable` conformance yourself locally in your own project code. You don't get the automatic generation of the `encode` function though so you would have to implement that yourself.

Unfortunately you will also have to do the same for every object `DetectDocumentTextResponse` references which is quite a number
```swift
extension Textract.DetectDocumentTextResponse: Encodable {
    func encode(to encoder: Encoder) {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(self.blocks, forKey: .blocks)
        try container.encode(self.detectDocumentTextModelVersion, forKey: .detectDocumentTextModelVersion)
        try container.encode(documentMetadata, forKey: .documentMetadata)
    }

    private enum CodingKeys: String, CodingKey {
        case blocks = "Blocks"
        case detectDocumentTextModelVersion = "DetectDocumentTextModelVersion"
        case documentMetadata = "DocumentMetadata"
    }
}
```

**Reply by DexterW** [reply]
> Adding `Encodable` conformance to the response shapes would add a lot of code to the libraries. The soto libs are already large enough. You can add the `Encodable` conformance yourself locally in your own project code. You don't get the automatic generation of the `encode` function though so you would have to implement that yourself.


Thanks for your response!  It seems to me that being able to cache AWS responses that cost $$ would add value to the library.  I'm happy to consider making a PR myself, but it sounds like you don't like the idea.  Is there perhaps a way to get access to the raw response JSON?  Then I'd be able to simply decode it into a `DetectDocumentTextResponse` repeatedly.  Thanks!

**Comment by adam-fowler** [comment]
You could add a middleware that saves the responses. eg
```swift
class CacheResponsesMiddleware: AWSServiceMiddleware {
    public func chain(response: AWSResponse, context: AWSMiddlewareContext) throws -> AWSResponse {
        guard case .json(let byteBuffer) = response.body else { return response }
        cacheByteBuffer(byteBuffer)
        return response
    }
}
```
You can then create a service using that middleware
```swift
let texExtract = TexExtract(client: ...).with(middlewares: [CacheResponsesMiddleware()])
```
